### ISSUE TITLE
Is this module even available anymore?

### ISSUE THREAD

**Issue by blackanvil23** [GEMakers/green-bean#12]
Can't find anything on the website, nor is google finding any source other than to that site. Is the module even still available?

**Comment by puddly** [comment]
Did you ever manage to source one?

**Comment by wied03** [comment]
I've had the same question. I managed to get a ConnectPlus but that's it.

**Comment by blackanvil23** [comment]
Nope.

**Comment by jallen2281** [comment]
yes, it is available, however, you need to compile as there are no prebuilt binaries. To do this and get the install to complete, first install these:
`sudo apt-get install libusb-1.0-0-dev libudev-dev 
`
and of course make sure you have git installed.

`sudo apt-get install git
`
Then install the module. 

`npm install green-bean
`
+ green-bean@0.1.1
added 67 packages from 47 contributors and audited 148 packages in 43.367s
found 0 vulnerabilities

**Comment by wied03** [comment]
@jallen2281 Do you have a link to purchase one?

**Comment by puddly** [comment]
@jallen2281 I think "adapter" may have been a better word to use in the title. The Node module only provides the computer to Green Bean interface, which is useless without the Green Bean to GE appliance protocol/the Green Bean adapter itself.

**Comment by jallen2281** [comment]
@wied03 @puddly Ah yes, my apologies. I purchased the module a while ago and it is no longer for sale. You might have luck searching ebay but otherwise you're out of luck. The node module is still available though.

**Comment by wied03** [comment]
@jallen2281 you might be able to make a few bucks by reverse engineering it and selling clones to us :)

**Comment by puddly** [comment]
If @jallen2281 (or someone else) is willing to lend me their module for a few days or use a logic analyzer and sniff its communications with a dishwasher or something, I can try to open source the protocol.

FirstBuild hasn't returned my calls or responded to my questions other than telling me the module is not for sale so I think the Green Bean hardware is very dead at this point.

**Comment by JonnyBoats** [comment]
Would it be worth contacting GE directly to see if they would be willing to release their protocol documentation to the public domain?

**Comment by jallen2281** [comment]
You would want to reach out to the firstbuild division, which is GE's maker
community division. They would likely have the needed schematics and board
files to manufacture a few. I would love to have a second board if someone
does decide to go through with it! But it has been said that they are unresponsive, so good luck! I hope someone is successful

**Comment by puddly** [comment]
Maybe they'll consider open sourcing it if more people ask them. It seems like a total of three people actually own these things so reverse engineering one is no easier than finding one...

The board itself should be easily replaceable with an off-the-shelf Arduino or other small development board so only the protocol is really necessary. Maybe it's enough to look at the more consumer-grade board that connects over WiFi?

**Comment by puddly** [comment]
I was never able to find a Green Bean so I've just started decoding the protocol from scratch after discovering that all the knobs and buttons on my washer actually communicate using the bus exposed by the RJ45 port. I've documented my initial attempt in [this repository](https://github.com/puddly/casserole). Let me know if you're interested in helping.

**Comment by puddly** [comment]
For anybody that's still interested, I finally managed to build a board that replaces the Green Bean entirely (with a few more interesting features).

I realized a little while ago that washers can't even be remotely started with a Green Bean because that functionality was only implemented by GE/FirstBuild for appliances used in a hackathon. You can only use the ConnectPlus, which my model doesn't even support.

But as it turns out, you can practically automate any GE appliance by just publishing messages to the bus simulating the buttons and knobs a person would turn. I'm now the proud owner of the first remotely-starting GUD24ESSMWW 😄

**Comment by jallen2281** [comment]
That's awesome, I'll buy 2!

**Comment by puddly** [comment]
@jallen2281 I wasn't really planning on making any PCBs but if enough people are interested, maybe in a few months (or much sooner if someone with actual PCB design experience helps out) and after I get a real Green Bean to test with so that I can get the protocol timing stuff right. Otherwise, you may find yourself replacing appliance innards because something got fried.

**Comment by jallen2281** [comment]
I have said experience. I'd be happy to help.

**Comment by puddly** [comment]
@jallen2281 thank you. I'll tag you in a couple of days in a GitHub issue on my project page when I've written up something concrete.

**Comment by ianbrennan** [comment]
I think I have 2 of the adapters somewhere in my house, I always wanted to write a Control4 automation driver for them but never had time.  Send me a message if you are interested in adapter I have.

**Comment by BrianHenryIE** [comment]
I bought the [GE Microwave](https://www.amazon.com/dp/B07D4821JB) and have spent a lot of time reverse engineering it. It uses GE Brillion, aka WiFi Connect, which I infer from the [Connect Plus WiFi module](https://www.h-mac.com/GE-PBX10W00Y0.html) using the same 8P8C (RJ45) connection must be using the same protocol to talk to the devices as Green Bean.

From digging around the [GE Kitchen iOS app](https://apps.apple.com/us/app/kitchen-ge-appliances/id992883749), I can see, once authenticated, it's using XMPP to communicate, and inside that XMPP it uses JSON dictionaries of hex address and values to send the commands and updates. 

e.g. Scroll to the bottom here to see the door open update being sent from the microwave to the iOS app: https://brianhenryie.s3.amazonaws.com/2019/door_opened.xmpp.log.html

So, I believe the Connect Plus WiFi module could be used in place of the Green Bean. I'm interested in getting my microwave hooked up to [Home Assistant](https://github.com/home-assistant/home-assistant). It requires your DHCP server returning a local address instead of the GE XMPP server, then running a pseudo XMPP server which understands the protocol used by the appliances which I assume are somewhat documented in this repo vis a vis the Green Bean commands. The non-documented commands could be determined by using a jailbroken iPhone and logging as buttons are pressed and as the UI updates in the Kitchen app, while also logging the XMPP.

**Comment by puddly** [comment]
@BrianHenryIE the appliances all seem to use the same protocol internally and externally via these plug-in devices, so it's just a matter of connectivity and the exposed interface (i.e. USB for the Green Bean, WiFi for the ConnectPlus, and eventually both + BT with my implementation).

Are you able to communicate with your microwave completely within your local network? I looked at the ConnectPlus a while ago but from what I recall, there was no plaintext traffic and my phone never communicated directly with it (despite being on the same network).

**Comment by puddly** [comment]
@ianbrennan I'm definitely interested in one of your Green Beans. I can't find any contact info in your GitHub profile so send me [an email](mailto:puddly@waiver.33mail.com) when you get a chance.

**Comment by BrianHenryIE** [comment]
It's certainly achievable. I used a [Python XMPP MITM script](https://github.com/iamultra/xmppmitm/blob/master/xmppmitm.py) which listens for the microwave to connect to it, supplies a self-signed certificate, then connects to the GE servers as though it's the microwave and passes the microwave's credentials to the server, from there eavesdropping on the normal microwave <-> app communication.

The iOS app uses certificate pinning which prevented using Charles proxy etc until I used a jailbroken phone, but there was no trouble with the microwave.

I'm only seeing your repo now. Good work!

**Comment by wied03** [comment]
@ianbrennan still have a green bean?

**Comment by JonnyBoats** [comment]
FYI I have a greanbean module if anyone wants information (photos etc) let me know and I will post it.

**Comment by jcastro14** [comment]
JonnyBoats, I'd be interested in seeing some detailed photos of the greenbean module.  A good detailed photo of the top and bottom would be awesome.  Pictures detailed enough to see chip labels, etc.

**Comment by mattgallion** [comment]
I know I'm rehashing an old thread, but I'm interested in this hardware as well. I do PCB design for a living so I'm interested in working on an open source design for this.

**Comment by mguaylam** [comment]
Look like something is cooking 🍳 : https://firstbuild.com/inventions/home-assistant-adapter/
They also hinted in one of the video that if there is popular demand, they could bring back the Green Bean.

**Comment by ianbrennan** [comment]
I have a green bean for sale if anybody is interested.

**Comment by Ron-N5HYH** [comment]
It looks like firstbuild.com is going to release new "GreenBean"  I would expect it to be something like either an Ether Net to internal Bus adaptor or an WIFI to Internal Buss Adaptor.  The latter could be easily built with an ESP32 a hand full of parts and some code to glue it all together.

**Comment by mguaylam** [comment]
I have made a post about the new module [here](https://community.home-assistant.io/t/home-assistant-adapter-for-ge-appliances).

**Comment by TiddlyWiddly** [comment]
I signed up, but didn't get an email confirmation. Is there somewhere I can purchase this?
